src/check.h

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CHECK_H
+#define BITCOIN_CHECK_H
+
+#include <cstdio>
+#include <cstdlib>
+
+// CHECK(expression)
+// =================
+// * If compiled with -DABORT_ON_FAILED_CHECK (set by --enable-debug):
+//   Evaluate expression and abort if expression is falsy.
+//
+// * If compiled without -DABORT_ON_FAILED_CHECK:
+//   Evaluate expression and continue execution.
+//
+// Example
+// =======
+// int main(void) {
+//     CHECK(IsFoo());
+//     ...
+// }
+//
+// If !IsFoo() and -DABORT_ON_FAILED_CHECK, then:
+//     filename.cpp:123: main: CHECK(IsFoo()) failed.
+//     Aborted
+// Otherwise the execution continues.
+
+#ifdef ABORT_ON_FAILED_CHECK
+#define CHECK_FAILURE(expression, file, line, func)       \
+    do {                                                  \
+        fprintf(stderr, "%s:%d: %s: CHECK(%s) failed.\n", \
+            file, line, func, expression);                \
+        std::abort();                                     \
+    } while (false)
+#else
+#define CHECK_FAILURE(expression, file, line, func) ((void)0)
+#endif
+
+#define CHECK(expression)                                             \
```

Comment:
This is missing an explanation when to use it. How is it different from `assert` or `std::logic_error`? Where should it be used in net or in consensus/validation code?

Reply:
Added additional documentation. Please re-review :-)